---
title: "PCA analysis of Gat201 RNA-seq dataset"
author: "Liz Hughes and Edward Wallace"
date: "16/12/2021"
output:
  html_document:
    toc: yes
    toc_depth: 2
  word_document:
    toc: yes
    toc_depth: '2'
  pdf_document:
    toc: yes
    toc_depth: '2'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE , warning=FALSE, message=FALSE)
```

# Summary

This document is a big-picture analysis of trends in the Gat201 RNA-seq dataset.

The first goal is a principal component analysis (PCA) to show the main axes of variation in the data.

# Load packages used for analysis

```{r load_packages, include=FALSE}

library(tidyr)
library(readr)
library(dplyr)
library(readxl)
library(stringr)
library(ggplot2)
library(cowplot)
library(biobroom)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(magrittr)
library(markdown)
library(forcats)
library(genefilter)
library(ggrepel)
library(extrafont)

theme_set(
  theme_cowplot(font_size = 12, font_family = "Arial") + 
    theme(strip.background = element_blank(),
          # panel.border = 
          #   element_rect(color = "grey90", 
          #                fill = NA, 
          #                linetype = 1, 
          #                size = 1),
          panel.grid.major = element_line(colour = "grey90", size = 0.5))
)

```


# Load count data and remove unwanted parts of column names

```{r load_count_data}
# note: we should insist that counts are loaded as integers
# and check the other data types
counts <- readr::read_tsv("../quantseqfwd_EH_050221/counts.txt",
                   comment = "#") %>% 
          dplyr::rename_with(str_remove_all, pattern = "_S[0-9]+_R1_001_aln.bam")

```


# Load sample sheet and format for DESeq2's requirements: readxl::read_excel

```{r load_sample_sheet}

samplesheet <- readxl::read_excel("../input_experiment/Gat201_samplesheet.xlsx") %>%
    magrittr::set_rownames(.$SampleID) %>%
    dplyr::mutate(GAT201 = 
                    forcats::fct_collapse(Strain, 
                                          "WT" = c("a","A"), 
                                          "del" = c("B","M")) %>% 
                    forcats::fct_relevel("del")
                  ) 

```

# Select counts

```{r select_counts_all}

counts_all <-
      dplyr::select(counts, samplesheet$SampleID) %>%
      magrittr::set_rownames(counts$Geneid)

```

# Construct the data object from the matrix of counts and the samplesheet information table

```{r dds_counts_all}

dds_all <- DESeqDataSetFromMatrix(countData = counts_all,
                                   colData = samplesheet,
                                   design = ~ Condition  + Time + GAT201)

```


# rlog transformation

```{r rlog_dds_all}

rld_all <- rlog(dds_all)
head(assay(rld_all))

```


## Calculate principal components of rlog

```{r calculate_pca_rlog}

## calculate the principal components sample-wise
pca_rlog <- prcomp(t(assay(rld_all)))

# combine pc projections to samples with sample sheet
pcdf_rlog <- bind_cols(
  as_tibble(colData(rld_all)) %>%
    mutate(GAT201_Cond = interaction(GAT201, Condition, sep = "_"),
           GAT201_Serum = forcats::fct_collapse(
             GAT201_Cond,
             "del_none"   = c("del_R", "del_Y"),
             "WT_none"   = c("WT_R", "WT_Y"),
             "del_Serum" = "del_RS",
             "WT_Serum"  = "WT_RS"
           ),
    ),
  as_tibble(pca_rlog$x)
)

propvar_rlog_df <- tibble(
  PC = seq.int(1L, ncol(pca_rlog$x) ),
  prop_var = pca_rlog$sdev^2/sum(pca_rlog$sdev^2)
)
```


# Plot PCA percentage variance explained to find how many PCs to pay attention to.

```{r plot_percentvar, fig.width = 3, fig.height = 2}
plot_percentvar_rlog <- 
  ggplot(data = propvar_rlog_df, 
         aes(x = PC, y = prop_var)) +
  geom_col(fill = "blue") +
  scale_x_continuous("principal component",
                     limits = c(0.4,10.6), 
                     # hack xlim to include the bars and a little space.
                     breaks = 1L:10L,
                     expand = c(0,0)) + 
  scale_y_continuous("prop. of variance", expand = c(0,0)) +
  theme(panel.grid.major = element_blank())

plot_percentvar_rlog
```



# Plot PC1 vs PC2 with default values

```{r plot_PC12_default}
ggplot(data = pcdf_rlog,
       aes(colour = factor(Time), 
           shape = GAT201_Cond)
       ) +
  geom_point(aes(x = PC1, y = PC2)) 
```

# Plot PC1 vs PC2 prioritising Time and GAT201_serum

```{r plot_PC12_fancy, fig.height=2, fig.width=4.25}
plot_PCA_Time_GC <- function(xpc = "PC1", ypc = "PC2",
                             data = pcdf_rlog) {
  # Plots PCA with default settings to highlight:
  #  Time as colour
  #  GAT201_Serum as shape, i.e. interaction of GAT201 and serum 
  #    (treating R and Y both as "no serum")
  # 
  # Default data input is pcdf_rlog, but it should work with others
  # as long as they have Time and GAT201_Serum as variables.
  
  ggplot(data = data,
         aes(colour = factor(Time), 
             shape = GAT201_Serum)
  ) +
    geom_vline(xintercept = 0, colour = "grey90", size = 0.5) +
    geom_hline(yintercept = 0, colour = "grey90", size = 0.5) +
    geom_point(aes_string(x = xpc, y = ypc)) +
    scale_shape_manual("Strain, Serum",
                       guide = guide_legend(ncol= 2),
                       values = c("del_none"   = 2,
                                  "WT_none"   = 1,
                                  "del_Serum" = 17,
                                  "WT_Serum"  = 16),
                       labels = c("del_none"   = "∆,  none",
                                  "WT_none"   = "WT, none",
                                  "del_Serum" = "∆,  + serum",
                                  "WT_Serum"  = "WT, + serum")
    ) +
    scale_colour_manual("Time (mins.)",
                        values = c("0" = "grey20",
                                   "30" = "#8c96c6",
                                   "120" = "#8856a7",
                                   "240" = "#810f7c")) +
    theme(axis.line = element_blank(), 
          panel.grid.major = element_blank(),
          panel.border = element_rect(colour= "grey50", size = 0.5))
  
}

plot_PC12_fancy <- 
  plot_PCA_Time_GC("PC1", "PC2")  +
  labs(x = paste0("PC1, ", 
                  round(propvar_rlog_df$prop_var[1] * 100),
                  "% of var."),
       y  = paste0("PC2, ", 
                   round(propvar_rlog_df$prop_var[2] * 100),
                   "% of var."))

plot_PC12_fancy
```

This plot shows that PC1 and PC2 are almost completely reporting on Time. Because PC1/2 collectively explain 80% of the variance, that establishes that Time is the biggest driver of variability.


```{r save_PC12_fancy}
ggsave("../results/PC12_plot_GAT201RNAseq.svg",
       plot_PC12_fancy, height = 2, width = 4.25)
```


## Make aligned plot of first 4 PCs

This is a draft of a multi-panel plot that could be a supplementary figure. 

```{r plot_PC1to4,fig.width = 7.5, fig.height = 7}
# Create plot panels but without legends
PC1vs2 <- plot_PCA_Time_GC("PC1", "PC2") + theme(legend.position = "none")
PC1vs3 <- plot_PCA_Time_GC("PC1", "PC3") + theme(legend.position = "none")
PC1vs4 <- plot_PCA_Time_GC("PC1", "PC4") + theme(legend.position = "none")
PC2vs3 <- plot_PCA_Time_GC("PC2", "PC3") + theme(legend.position = "none")
PC2vs4 <- plot_PCA_Time_GC("PC2", "PC4") + theme(legend.position = "none")
PC3vs4 <- plot_PCA_Time_GC("PC3", "PC4") + theme(legend.position = "none")

# get the legend
PC_legend_Time_GC <- get_legend( 
  plot_PCA_Time_GC() + 
    theme(legend.position = c(0.05, 0.6),
      legend.box.margin = margin(4, 5, 4, 4), 
          legend.box.background = element_rect(colour= "grey50", size = 0.25))
  )

# plot those aligned in a grid

plot_grid(PC1vs2, plot_percentvar_rlog, NULL,   
          PC1vs3, PC2vs3, PC_legend_Time_GC,
          PC1vs4, PC2vs4, PC3vs4, 
          ncol = 3,
          align = "hv")

ggsave("../results/PCmultipanel_plot_GAT201RNAseq.png",
       plot_PC12_fancy, height = 7, width = 7.5)
```

This plot shows that PC3 reports on Gat201 status,  vs WT. 

It is unclear what PC4 means.

## PC5 somewhat reports on serum at the later time points

I found the clearest report in PC2 vs PC5, because PC2 lays out the times in the expected order.

Do we need to include this plot panel?

```{r plot_PC25_fancy}
plot_PCA_Time_GC("PC2", "PC5")
```


## To do

- decide which plots to include in 
- try again with vst and see if it's different in any important way
