---
title: "Serum is not the dominant driver of gene expression"
author: "Liz Hughes and Edward Wallace"
date: "2023-04-24"
output:
  html_document:
    toc: yes
    toc_depth: 2
  word_document:
    toc: yes
    toc_depth: '2'
  pdf_document:
    toc: yes
    toc_depth: '2'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE , warning=FALSE, message=FALSE)
```

# Summary

This document is an analysis of the Gat201 RNA-seq dataset, focusing on differences that depend on serum. This means that we treat the two WT strains as the same, and the two ∆gat201 strains as the same.

This makes figure panels:

- heatmap of log2FC for all (or most) genes, showing overall variability, strong dependence on time, weak dependence on serum.
- serum-dependent differential gene expression, counts per strain and timepoint

These will then be combined into a multipanel figure, planned as figure 3 of the manuscript.

This also outputs serum-dependent differential expressed data frames and genelists:

- DEGdf_WT_30_vsSerum_down2x_FDR5_ids.txt
- DEGdf_WT_30_vsSerum_down2x_FDR5.txt
- DEGdf_WT_30_vsSerum_up2x_FDR5_ids.txt
- DEGdf_WT_30_vsSerum_up2x_FDR5.txt
- DEGdf_WT_120_vsSerum_down2x_FDR5_ids.txt
- DEGdf_WT_120_vsSerum_down2x_FDR5.txt
- DEGdf_WT_120_vsSerum_up2x_FDR5_ids.txt
- DEGdf_WT_120_vsSerum_up2x_FDR5.txt
- DEGdf_WT_240_vsSerum_down2x_FDR5_ids.txt
- DEGdf_WT_240_vsSerum_down2x_FDR5.txt
- DEGdf_WT_240_vsSerum_up2x_FDR5_ids.txt
- DEGdf_WT_240_vsSerum_up2x_FDR5.txt
- DEGdf_delGAT201_240_vsSerum_up2x_FDR5.txt
- DEGdf_delGAT201_30_vsSerum_down2x_FDR5_ids.txt
- DEGdf_delGAT201_30_vsSerum_down2x_FDR5.txt
- DEGdf_delGAT201_30_vsSerum_up2x_FDR5_ids.txt
- DEGdf_delGAT201_30_vsSerum_up2x_FDR5.txt
- DEGdf_delGAT201_120_vsSerum_down2x_FDR5_ids.txt
- DEGdf_delGAT201_120_vsSerum_down2x_FDR5.txt
- DEGdf_delGAT201_120_vsSerum_up2x_FDR5_ids.txt
- DEGdf_delGAT201_120_vsSerum_up2x_FDR5.txt
- DEGdf_delGAT201_240_vsSerum_down2x_FDR5_ids.txt
- DEGdf_delGAT201_240_vsSerum_down2x_FDR5.txt
- DEGdf_delGAT201_240_vsSerum_up2x_FDR5_ids.txt


# Setup: load packages, load data, clean and prep data


## Load packages used for analysis

```{r load_packages, include=FALSE}
library(tidyr)
library(readr)
library(dplyr)
library(readxl)
library(stringr)
library(ggplot2)
library(cowplot)
library(biobroom)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(magrittr)
library(forcats)
library(genefilter)
library(ggrepel)
library(ggdendro)
library(vegan)
library(extrafont)

theme_set(
  theme_cowplot(font_size = 12, font_family = "Arial") + 
    theme(strip.background = element_blank(),
          # panel.border = 
          #   element_rect(color = "grey90", 
          #                fill = NA, 
          #                linetype = 1, 
          #                size = 1),
          panel.grid.major = element_line(colour = "grey90", size = 0.5),
          plot.margin = margin(0,0,0,0))
)

```

```{r, include = FALSE}
write_DEGdflist <- function(DEGdf, datadir = "../results/", digits = 5,
                            write_df = TRUE, write_list = TRUE) {
  # get the name of DEG
  DEGdfname <- deparse(substitute(DEGdf))
  # Writes a DEG dataframe and genelist
  if(write_df) {
    readr::write_tsv(x = DEGdf %>%
                       mutate_if(is.numeric, signif, digits = digits),
                     file = paste0(datadir, DEGdfname, ".txt") )
  }
  if(write_list) {
    writeLines(text = DEGdf %>%
                 pull(gene),
               con = paste0(datadir, DEGdfname, "_ids.txt") )
  }
}
```



## Load count data and remove unwanted parts of column names

```{r load_count_data}
# note: we should insist that counts are loaded as integers
# and check the other data types
counts <- readr::read_tsv("../quantseqfwd_EH_050221/counts.txt",
                   comment = "#") %>% 
          dplyr::rename_with(str_remove_all, pattern = "_S[0-9]+_R1_001_aln.bam")

```


## Load sample sheet and format for DESeq2's requirements: readxl::read_excel

Here we choose how the Strain labels behave by manipulating the factors. This creates a new variable `GAT201` whose levels are `del` and `WT`.

Similarly we create factors with the desired order for combinations of

- `Condition_Time`
- `GAT201_Time`
- `GCT` short for `GAT201_Condition_Time`, so it groups four biological replicates together.


```{r load_sample_sheet}

Condition_Time_levels <- c("Y_0", 
                           "R_30", "R_120", "R_240",
                           "RS_30", "RS_120", "RS_240")

# GAT201_Time
GAT201_Time_levels <- c("WT_0", "WT_30", "WT_120", "WT_240",
                        "del_0", "del_30", "del_120", "del_240")

## GAT201_Time, for vertical printing
GAT201_Time_levels_vprint <- 
  GAT201_Time_levels %>%
  stringr::str_replace("del","∆") %>%
  stringr::str_replace("_","\n")

# GAT201_Condition_Time
GCT_levels <- c("WT_Y_0", 
                "WT_R_30", "WT_RS_30", 
                "WT_R_120", "WT_RS_120", 
                "WT_R_240", "WT_RS_240",
                "del_Y_0", 
                "del_R_30", "del_RS_30", 
                "del_R_120", "del_RS_120", 
                "del_R_240", "del_RS_240")

samplesheet <- readxl::read_excel("../input_experiment/Gat201_samplesheet.xlsx") %>%
    magrittr::set_rownames(.$SampleID) %>%
    dplyr::mutate(GAT201 = 
                    forcats::fct_collapse(Strain, 
                                          "WT" = c("a","A"), 
                                          "del" = c("B","M")) %>% 
                    forcats::fct_relevel("del"),
                  Timepoint = factor(Timepoint),
                  Serum = forcats::fct_collapse(Condition, 
                                          "Yes" = c("RS"), 
                                          "No" = c("R")),
                  Condition_Time = factor(
                    paste(Condition, Time, sep = "_"),
                    levels = Condition_Time_levels
                  ),
                  GAT201_Time =  factor(
                    paste(GAT201, Time, sep = "_"),
                    levels = GAT201_Time_levels
                  ),
                  GCT = factor(
                    paste(GAT201, Condition, Time, sep = "_"),
                    levels = GCT_levels
                  )
    )

```

## Select counts

```{r select_counts_all, warning = FALSE}

counts_all <-
      dplyr::select(counts, samplesheet$SampleID) %>%
      magrittr::set_rownames(counts$Geneid)

```

# Normalised gene expression analysis with DESeq2

## Run DESeq2 analysis dependent on combination of GAT201, Condition, and Time

This first uses the matrix of counts and the samplesheet information table to construct the DESeq2 data object, then immediately runs the analysis.

The design formula `~ GCT - 1` ensures that every combination of GAT201, Condition and Time has a coefficient, where we remove the intercept term by including `- 1`. Removing the intercept term means that we have a coefficient for the first value of GCT, i.e. `WT_Y_0`.

```{r deseq_GAT201_Condition_Time}

dds_all <- DESeqDataSetFromMatrix(
    countData = counts_all,
    colData = samplesheet,
    design = ~ GCT - 1 ) %>%
  DESeq()

dds_all
resultsNames(dds_all)
```

## Check distribution of counts and baseMean

```{r count_distribution}
count_stats_bygene <- 
  counts_all %>%
  as.matrix() %>%
  tibble(gene = rownames(.),
         count_min = rowMins(.),
         count_mean = rowMeans(.),
         count_max = rowMaxs(.)) %>%
  select( - ".") # removes pipe behaviour leading to inclusion of "."

genelist_mincount_10 <- 
  count_stats_bygene %>%
  filter(count_min >= 10) %>%
  pull(gene)

ggplot(data = count_stats_bygene) +
  geom_density(aes(x=count_min + 1, colour = "min"), 
               size = 1, kernel = "rectangular") +
  geom_density(aes(x=count_max + 1, colour = "max"), 
               size = 1, kernel = "rectangular") +
  geom_density(aes(x=count_mean + 1, colour = "mean"), 
               size = 1, kernel = "rectangular") +
  scale_colour_manual(values = c("min" = "lightblue",
                                 "mean" = "grey20",
                                 "max" = "pink")) + 
  scale_x_log10("log10(count per gene + 1)")
```


## Select GCT-dependent log2FC at each timepoint

```{r tidy_deseq_GCT}
# Create a tidy data frame that contains only the GAT201-dependent log2FC
# in a helpful format for plotting
deseq_df_GCT <- 
  biobroom::tidy.DESeqDataSet(dds_all) %>%
  mutate(GCT = term %>%
           stringr::str_remove("GCT") %>%
           factor(levels = GCT_levels)) %>%
  select(GCT, gene, baseMean, log2estimate = estimate) %>%
  group_by(gene) %>%
  mutate(log2FC = log2estimate - mean(log2estimate)) %>%
  ungroup()

# restrict to genes with mincount 10
deseq_df_GCT_mincount_10 <- 
  deseq_df_GCT %>%
  filter(gene %in% genelist_mincount_10)

# check we retained all the levels
unique(deseq_df_GCT$GCT)
```

## Sanity-check: Did we get the log2FC right for a favourite gene?

```{r print_log2FC_6917}
deseq_df_GCT %>%
  filter(gene == "CNAG_06917")
```

```{r plot_log2FC_6917}
ggplot(data = deseq_df_GCT %>%
  filter(gene %in% c("CNAG_00483", "CNAG_00121", "CNAG_06917", "CNAG_01551")),
  aes(x = GCT, fill = log2FC, y = gene)) +
  geom_tile() +
  scale_fill_gradient2(low = "cyan", mid = "black", high = "yellow") +
  theme(axis.text.x = element_text(angle = -45, hjust = 0.1),
        panel.grid.major = element_blank(),
        axis.line = element_blank())
```

## Plot log2FC for genes with mincount 10, with no clustering

```{r plot_log2FC_m10}
ggplot(data = deseq_df_GCT_mincount_10,
  aes(x = GCT, fill = log2FC, y = gene)) +
  geom_tile() +
  scale_fill_gradient2(low = "cyan", mid = "black", high = "yellow",
                       limits = c(-3.5, 3.5), oob = scales::squish) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0.1),
        panel.grid.major = element_blank(),
        axis.line = element_blank(),
        axis.text.y = element_blank(), axis.ticks.y = element_blank())
```

## Calculate gene expression hierarchical clusters

Here we restrict to genes with at least 10 counts in each sample.
This list was calculated earlier as `genelist_mincount_10`. 

This avoids a problem with numeric errors in calculations for all enes, that probably came from genes with low/zero counts.

```{r hclust_genes}
log2FC_GCT_m10_wide <-
  deseq_df_GCT_mincount_10  %>%
  select(gene,GCT, log2FC) %>%
  pivot_wider(id_cols = gene, names_from = GCT, values_from = log2FC) %>%
  select(c("gene", GCT_levels))

# order genes according to mean value at 0min, as far as clustering allows
preferred_gene_ordering <- 
  log2FC_GCT_m10_wide %>% 
  dplyr::mutate(Y_0_mean = ( WT_Y_0 + del_Y_0 ) / 2 ) %>%
  dplyr::pull(Y_0_mean)

distcort <- function(x) {
  as.dist(1 - cor(t(x)))
}

hclust_log2FC <- 
  log2FC_GCT_m10_wide %>%
  select(-gene) %>%
  as.matrix() %>%
  set_rownames(log2FC_GCT_m10_wide$gene) %>%
  # head(n= 100) %>%
  distcort() %>%
  hclust(method = "average") %>%
  reorder(wts = preferred_gene_ordering)
```
## Plot clustering dendrogram by itself

```{r plot_dendrogram_log2FC_m10}
plot_dendrogram_log2FC_m10 <- 
  ggdendrogram(hclust_log2FC,
               rotate = TRUE,
               labels = FALSE) + 
  theme_void() +
  scale_y_reverse() +
  coord_flip(expand = c(0,0), clip = "off")
plot_dendrogram_log2FC_m10
```

## Plot the log2 fold-change in same order as clustering dendrogram

```{r plot_log2FC_m10_dorder}

deseq_df_GCT_mincount_10_dorder <- 
  deseq_df_GCT_mincount_10 %>%
  mutate(gene_dorder = 
           factor(gene,
                  levels = log2FC_GCT_m10_wide$gene[hclust_log2FC$order],
                  ordered = TRUE)
         )

plot_log2FC_m10_dorder <- 
  ggplot(data = deseq_df_GCT_mincount_10_dorder,
         aes(x = GCT, fill = log2FC, y = gene_dorder)) +
  geom_tile() +
  scale_fill_gradient2(low = "cyan", mid = "black", high = "yellow",
                       limits = c(-3.5, 3.5), oob = scales::squish) +
  coord_cartesian(expand = c(0,0), clip = "off") + 
  theme(axis.text.x = element_text(angle = -45, hjust = 0.1),
        panel.grid.major = element_blank(),
        axis.line = element_blank(),
        axis.text.y = element_blank(), axis.ticks.y = element_blank()) +
  labs(x = "GAT201, Media, Time",
       y = "Gene, ordered by clustering")

plot_log2FC_m10_dorder
```

## Make a nicer x-axis for sample labels

This section programs up x-axis labels split up by GAT201, Time, Serum

It's ugly code to get a pretty result.

```{r nice_GCT_labels, fig.height = 1, fig.width = 5}
GCT_df <- 
  tibble(GCT = as_factor(GCT_levels),
                 x = as.numeric(GCT)) %>%
  separate(GCT, c("GAT201","Condition","Time"), remove = FALSE) %>%
  mutate(Time = as.integer(Time),
         Serum = factor(Condition, 
                        levels = c("R","RS"),
                        labels = c("−", "+")),
         GAT201 = GAT201 %>% 
           str_replace("del", "gat201∆") %>% 
           str_replace( "WT", "GAT201"),
         GAT201Serum = paste(GAT201, Condition, sep = "_") %>% 
           str_replace("RS", "+") %>%
           str_replace("R|Y", "−") %>%
           as_factor() 
         )

GCT_time_df <- GCT_df %>%
  group_by(GAT201,Time) %>%
  summarise(xmean = mean(x), xmin = min(x), xmax = max(x))

GCT_GAT201_df <- GCT_df %>%
  group_by(GAT201) %>%
  summarise(xmean = mean(x), xmin = min(x), xmax = max(x))

plot_nice_GCT_labels <- 
  ggplot(data = GCT_df) +
  geom_text(aes(x = x, label = Serum), y = 3) + 
  geom_text(data = GCT_time_df, aes(x = xmean, label = Time), y = 2) + 
  geom_segment(data = GCT_time_df, 
               aes(x = xmin, xend = xmax), y = 2.5, yend = 2.5) +
  geom_text(data = GCT_GAT201_df, aes(x = xmean, label = GAT201), y = 1) +
  geom_segment(data = GCT_GAT201_df, 
               aes(x = xmin, xend = xmax), y = 1.5, yend = 1.5) +
  scale_y_continuous(limits = c(0.5,3.5),
                     breaks = c(1,2,3),
                     labels = c("Strain", "Time (mins)", "Serum")) + 
  scale_x_continuous(limits = c(0.5, max(GCT_df$x + 0.5))) + 
  coord_cartesian(expand = c(0,0), clip = "off") + 
  theme_nothing() + 
  theme(axis.text.y = element_text())

plot_nice_GCT_labels

# the next bit plots the y-axis labels as their own ggplot, right-aligned
# so that we can get everything to align nicely in the plot_grid later
plot_nice_GCT_yaxis <- 
  ggplot(data = tibble(y = c(1,2,3),
                     label = c("Strain", "Time", "Serum")),
         aes(y=y, label = label)) +
  geom_text(hjust = 1, x = 1) +
  coord_cartesian(xlim = c(0, 1), ylim = c(0.5, 3.5),
                  expand = c(0,0), clip = "off")  + 
  theme_nothing() 
```


# Combined plot of dendrogram and heatmap

This combines the previous plot panels into a single plot, showing the dendrogram aligned with the clustered heatmap, and with nice x-axis labels.

```{r combine_dendrogram_heatmap, fig.height = 5, fig.width = 5}
plot_dh_nolegend <- 
  plot_grid(plot_dendrogram_log2FC_m10,
          plot_log2FC_m10_dorder +
            theme(axis.title.y = element_blank(),
                  axis.title.x = element_blank(),
                  axis.text.x = element_blank(),
                  axis.ticks.x = element_blank(),
                  legend.position = "none"),
          plot_nice_GCT_yaxis,
          plot_nice_GCT_labels + theme(axis.text.y = element_blank()),
          ncol = 2,
          align = "hv",
          rel_widths = c(1,5),
          rel_heights = c(7,1))

plot_grid(get_legend(plot_log2FC_m10_dorder + 
                       labs(fill = "mRNA log2 fold-change") +
                       theme(legend.box.margin = margin(0, 0, 0, 2, 
                                                        unit = "cm"), 
                             legend.position = "top")),
          plot_dh_nolegend,
          ncol = 1,
          rel_heights = c(1,7))

ggsave(filename = "../results/clustered_heatmap_GAT201SerumTime.svg",
       width = 5, height = 5)
```

### Highlight select genes on y-axis.

```{r highlight_select}

select_genes <- c("CNAG_03012", "CNAG_06443", "CNAG_01551", "CNAG_06917", "CNAG_06125", "CNAG_06762", "CNAG_00979")

select_genes_dorder <- 
  tibble( gene = log2FC_GCT_m10_wide$gene,
          gene_dorder = factor(log2FC_GCT_m10_wide$gene,
                  levels = log2FC_GCT_m10_wide$gene[hclust_log2FC$order],
                  ordered = TRUE),
          y = as.integer(gene_dorder)) %>%
  dplyr::filter( gene %in% select_genes )

plot_select_genes_dorder <- 
  ggplot(data = select_genes_dorder)  +
  # geom_segment(colour = "grey50", 
  #              x = 0, xend = 0,
  #              y = 1, yend = nrow(log2FC_GCT_m10_wide)) +
  geom_text(aes(y = y, label = gene),
            x = 0, size = 2) +
  coord_cartesian(xlim = c(-0.5,0.5), 
                  ylim = c(1, nrow(log2FC_GCT_m10_wide)),
                  expand = c(0,0), clip = "off") +
  theme_nothing()
  
plot_grid(plot_dendrogram_log2FC_m10,
          plot_log2FC_m10_dorder +
            theme(axis.title.y = element_blank(),
                  axis.title.x = element_blank(),
                  axis.text.x = element_blank(),
                  axis.ticks.x = element_blank(),
                  legend.position = "none"),
          plot_select_genes_dorder,
          plot_nice_GCT_yaxis,
          plot_nice_GCT_labels + theme(axis.text.y = element_blank()),
          NULL,
          ncol = 3,
          align = "hv",
          rel_widths = c(1,5,0.6),
          rel_heights = c(7,1))

ggsave(filename = "../results/clustered_heatmap_GAT201SerumTime_plus.svg",
       width = 5.5, height = 5)
```


### Extract and plot clusters

```{r extract_cluster_functions}
  
make_clustered_df <- function(k = 2, 
                              names_df = log2FC_GCT_m10_wide,
                              hclust_results = hclust_log2FC) {
      tibble(gene = names_df$gene,
         gene_dorder = 
           factor(gene,
                  levels = names_df$gene[hclust_results$order],
                  ordered = TRUE),
         y = as.integer(gene_dorder),
         cluster = hclust_results %>% 
           cutree(k = k) %>% 
           factor())
}

get_cluster_from_clustered_df <- function(gene_bait, 
                                          clustered_df,
                                          output = c("list", "df")) {
  # find cluster containing the gene_bait (gene of interest)
  cluster_number <- clustered_df %>%
    dplyr::filter(gene == gene_bait) %>%
    dplyr::pull(cluster)
  # return all of that cluster
  clustered_df_filtered <- 
    clustered_df %>%
    dplyr::filter(cluster == cluster_number)
  if (output == "df") {
    return(clustered_df_filtered)
  } else if (output == "list") {
    return(clustered_df_filtered$gene)
  } 
}

make_cluster_with_gene <- function(gene_bait,
                                   k = 2, 
                                   names_df = log2FC_GCT_m10_wide,
                                   hclust_results = hclust_log2FC,
                                   output = c("list", "df")) {
  clustered_df <- 
    make_clustered_df(k = k, 
                      names_df = names_df,
                      hclust_results = hclust_results)
  get_cluster_from_clustered_df(gene_bait, clustered_df)
}

plot_clustered_genelist <- 
  function(genelist, data = deseq_df_GCT_mincount_10_dorder) {
    ggplot(data = data %>% 
             dplyr::filter(gene %in% genelist),
           aes(x = GCT, fill = log2FC, y = gene_dorder)) +
      geom_tile() +
      scale_fill_gradient2(low = "cyan", mid = "black", high = "yellow",
                           limits = c(-3, 3), oob = scales::squish) +
      scale_y_discrete(breaks = genelist#, labels = names(genelist)
                       ) +
      coord_cartesian(expand = c(0,0), clip = "off") + 
      theme(axis.text.x = element_text(angle = -45, hjust = 0.1),
            panel.grid.major = element_blank())
    
  }
```

```{r extract_clusters}
genecluster_CQS1 <- make_cluster_with_gene(gene = "CNAG_03012", k = 100)

plot_genecluster_CQS1 <- plot_clustered_genelist(genecluster_CQS1)


genecluster_SSA1 <- make_cluster_with_gene(gene = "CNAG_06443", k = 100)

genecluster_TEF1_k100 <- make_cluster_with_gene(gene = "CNAG_06125", k = 100)
all.equal(genecluster_SSA1,genecluster_TEF1_k100)

genecluster_TEF1 <- make_cluster_with_gene(gene = "CNAG_06125", k = 500)
genecluster_CTR4 <- make_cluster_with_gene(gene = "CNAG_00979", k = 500)
all.equal(genecluster_TEF1,genecluster_CTR4)

plot_genecluster_TEF1 <- plot_clustered_genelist(genecluster_TEF1)


genecluster_TSA3 <- make_cluster_with_gene(gene = "CNAG_06917", k = 100)

plot_genecluster_TSA3 <- plot_clustered_genelist(genecluster_TSA3)


genecluster_GAT201 <- make_cluster_with_gene(gene = "CNAG_01551", k = 100)

plot_genecluster_GAT201 <- plot_clustered_genelist(genecluster_GAT201)


```

```{r collect_clusters}
hclust_colour_df_grouped <-
  tibble(gene = log2FC_GCT_m10_wide$gene,
         gene_dorder =
           factor(gene,
                  levels = log2FC_GCT_m10_wide$gene[hclust_log2FC$order],
                  ordered = TRUE),
         y = as.integer(gene_dorder),
         cluster_select = fct_collapse(gene,
                                       CQS1   = genecluster_CQS1,
                                       # HSP    = genecluster_SSA1,
                                       GAT201 = genecluster_GAT201,
                                       TSA3   = genecluster_TSA3,
                                       RPs    = genecluster_TEF1,
                                       other_level = "other") %>%
           forcats::fct_recode(NULL = "other") 
         ) %>%
  group_by(cluster_select) %>%
  summarise(n = n(), ymin = min(y), ymax = max(y), ymean = mean(y))
```

```{r plot_clusters}
# FIX ME
plot_cluster_select <- 
  ggplot(data = hclust_colour_df_grouped %>%
           dplyr::filter(!is.na(cluster_select))) +
  geom_rect(aes(ymin = ymin, ymax = ymax
                , fill = cluster_select
                ),
            # fill = "grey50",
            xmin = -0.1, xmax = 0) +
  geom_text(aes(y = ymean,
                colour = cluster_select,
                label = cluster_select),
            x = 0.1, hjust = 0) +
  coord_cartesian(xlim = c(-0.1,1), 
                  ylim = c(min(hclust_colour_df_grouped$ymin),
                           max(hclust_colour_df_grouped$ymax)),
                  expand = c(0,0), clip = "off") +
  scale_colour_brewer(type = "qual") + scale_fill_brewer(type = "qual") +
  theme_nothing()
  
plot_cluster_select
```


## Plot example genes for main figure

```{r plot_log2FC_line_faceted_6select, fig.height = 6, fig.width = 3}

plot_mygenes_facet <- function(mygenes,
                               mygenenames, 
                               deseq_df = deseq_df_GCT,
                               meta_df = GCT_df,
                               facet_option = c("stack_free", "grid_fixed")) {
  mygenes_df <- 
    tibble(gene  = mygenes %>% as_factor(), 
           name = mygenenames %>% as_factor()) %>%
    dplyr::left_join(deseq_df) %>%
    dplyr::left_join(meta_df, by = "GCT")
  
    plot_out <- 
  ggplot(data = mygenes_df, 
         aes(x = Time, y = log2FC,
             colour = GAT201Serum, group = GAT201Serum,
             linetype = GAT201Serum)) +
    geom_hline(yintercept = 0, size = 0.5, colour = "grey90") +
    geom_line(size = 1) +
    scale_y_continuous("mRNA log2 fold-change",
                       # limits = ylimits,
                       breaks = seq(-4,4),
                       expand = c(0.01,0.2)) +
    scale_x_continuous("Time (mins.)", 
                       breaks = c(0,30,120,240),
                       expand = c(0.01,0.01)) + 
    scale_colour_manual(
      "GAT201, Serum",
      values = c("GAT201_−" = "grey30",
                 "GAT201_+" = "grey10",
                 "gat201∆_−" = "red2",
                 "gat201∆_+" = "red4")) +
    scale_linetype_manual(
      "GAT201, Serum",
      values = c("GAT201_−" = "21",
                 "GAT201_+" = "solid",
                 "gat201∆_−" = "21",
                 "gat201∆_+" = "solid")) +
    guides(colour = guide_legend(
      title.position = "top",
      ncol = 2)) +
    panel_border(color = "black", size = 0.5) +
    theme(strip.text.y = element_text(angle = 0),
          legend.position = "top",
          plot.margin = unit(c(0.1, 0, 0.1, 0), "in"),
          panel.grid.major.x = element_blank()) 
    if (facet_option == "stack_free") {
    plot_out <- plot_out + 
      facet_grid(name~., scales = "free", space = "free")
  } else if (facet_option == "wrap_fixed") {
    plot_out <- plot_out + 
      facet_wrap(~ name)
  }
  return(plot_out)
}

# plot_mygenes_facet(
#   mygenes =  c("CNAG_03012", "CNAG_06443", "CNAG_01551", "CNAG_06917", "CNAG_06125"),
#   mygenenames = c("CQS1", "SSA1", "GAT201", "TSA3", "TEF1")
# )

plot_mygenes_facet(
  mygenes =  c("CNAG_06917", "CNAG_03012", "CNAG_01551", "CNAG_06443", "CNAG_00979", "CNAG_06125"),
  mygenenames = c("TSA3", "CQS1", "GAT201", "SSA1", "CTR4", "TEF1")
)

ggsave(filename = "../results/log2FC_line_multipanel_6select.svg",
       width = 2.5, height = 5.8)
```

```{r plot_log2FC_line_faceted_Gat201dep}
plot_mygenes_facet(
  mygenes =  c("CNAG_06762", "CNAG_05835", "CNAG_00979", "CNAG_02775", "CNAG_00795"),
  mygenenames = c("GAT204", "LIV3", "CTR4", "BIM1", "CFL1")
)

ggsave(filename = "../results/log2FC_line_multipanel_Gat201dep.png",
       width = 2.5, height = 5)
```


```{r plot_log2FC_line_faceted_30genes, fig.height = 9, fig.width = 7.5}

my25genes <- read_tsv("../input_annotation/CryptoWakeup_30genes.txt")

plot_mygenes_facet(
  mygenes =  my25genes$GeneID,
  mygenenames = my25genes$GeneName,
  facet_option = "wrap_fixed"
)

ggsave(filename = "../results/log2FC_line_30genes.png",
       width = 7.5, height = 9.5)
```



# Serum-dependent gene expression analysis with DESeq2

## Run DESeq2 analysis dependent on Condition/Time and GAT201

This first uses the matrix of counts and the samplesheet information table to construct the DESeq2 data object, then immediately runs the analysis.

The design formula `~ Condition_Time + GAT201:Condition_Time - 1` ensures that every combination of Condition and Time has a coefficient, where we remove the intercept term by including `- 1`. Then the interaction term `GAT201:Condition_Time` gives the effect of `GAT201` separately at each combination of Condition and Time, which we are interested in.

```{r deseq_serum}

samplesheet_RPMIboth <- 
  dplyr::filter(samplesheet, Time > 0)

counts_RPMIboth <-
      dplyr::select(counts, samplesheet_RPMIboth$SampleID) %>%
      magrittr::set_rownames(counts$Geneid)

dds_serum <- DESeqDataSetFromMatrix(
    countData = counts_RPMIboth,
    colData = samplesheet_RPMIboth,
    design = ~ GAT201_Time + Serum:GAT201_Time - 1 ) %>%
  DESeq()

dds_serum
resultsNames(dds_serum)
```

## Select serum-dependent log2FC at each timepoint

```{r tidy_deseq_serum}
coefnames_serum <- c("GAT201_TimeWT_30.SerumYes", "GAT201_TimeWT_120.SerumYes",
                      "GAT201_TimeWT_240.SerumYes",  
                      "GAT201_Timedel_30.SerumYes", "GAT201_Timedel_120.SerumYes",
                      "GAT201_Timedel_240.SerumYes")

# Create a tidy data frame that contains only the GAT201-dependent log2FC
# in a helpful format for plotting
deseq_df_GT_vsserum <- 
  biobroom::tidy.DESeqDataSet(dds_serum) %>%
  dplyr::filter(term %in% coefnames_serum) %>%
  mutate(GAT201_Time = term %>%
           stringr::str_remove("GAT201_Time") %>%
           stringr::str_remove(".SerumYes") %>%
           factor(levels = GAT201_Time_levels)) %>%
  select(GAT201_Time, gene, baseMean, log2FC = estimate, stderror, padj = p.adjusted)

# check we retained all the levels
unique(deseq_df_GT_vsserum$GAT201_Time)
```

## Count the number of DEGs up and down at each Condition and Time

```{r countDEGs_GT_vsserum, fig.width = 6, fig.height = 3}
countDEGs_GT_vsserum <- 
  deseq_df_GT_vsserum %>%
  group_by(GAT201_Time) %>%
  summarise(Up   = sum(log2FC > 1  & padj < 0.05, na.rm = TRUE),
            Down = sum(log2FC < -1 & padj < 0.05, na.rm = TRUE))

## prototype plot; not run but included as an example.
# ggplot(data = countDEGs_CT_vsGAT201,
#        aes(x = GAT201_Time)) +
#   geom_col(aes(y = Up))

countDEGs_GT_vsserum_long <- 
  countDEGs_GT_vsserum %>%
  pivot_longer(cols = c(Up,Down),
               names_to = "Direction",
               values_to = "DEG_count")

barplot_countDEGs_GT_vsserum <-
  ggplot(data = countDEGs_GT_vsserum_long,
         aes(x = GAT201_Time, y = DEG_count, fill = Direction)) +
  geom_col(position = "dodge") +
  scale_fill_manual("Serum vs no Serum\nin RPMI",
                    values = c("Down" = "darkred",
                               "Up" = "darkblue")) +
  scale_y_continuous("gene count\n2x DE, 5% FDR", expand = c(0,0)) +
  scale_x_discrete("GAT201 status, Time (mins.)", 
                   breaks = GAT201_Time_levels, 
                   labels = GAT201_Time_levels_vprint) +
  theme(panel.grid.major.x = element_blank())

barplot_countDEGs_GT_vsserum
```


## Calculate serum-dependent DEG lists for both strains at each timepoint

Write output to tab-separated txt files, in order of log2 fold-change as calculated above.
Both as DEG dfs including log2 fold-change and p-values, also as lists of gene Ids only.


First for wild-type cells

```{r DEGs_vsSerum_WT}
DEGdf_WT_30_vsSerum_up2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "WT_30",
                log2FC > 1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(desc(log2FC))

write_DEGdflist(DEGdf_WT_30_vsSerum_up2x_FDR5)

DEGdf_WT_30_vsSerum_down2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "WT_30",
                log2FC < -1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(log2FC)

write_DEGdflist(DEGdf_WT_30_vsSerum_down2x_FDR5)


DEGdf_WT_120_vsSerum_up2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "WT_120",
                log2FC > 1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(desc(log2FC))

write_DEGdflist(DEGdf_WT_120_vsSerum_up2x_FDR5)

DEGdf_WT_120_vsSerum_down2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "WT_120",
                log2FC < -1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(log2FC)

write_DEGdflist(DEGdf_WT_120_vsSerum_down2x_FDR5)


DEGdf_WT_240_vsSerum_up2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "WT_240",
                log2FC > 1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(desc(log2FC))

write_DEGdflist(DEGdf_WT_240_vsSerum_up2x_FDR5)

DEGdf_WT_240_vsSerum_down2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "WT_240",
                log2FC < -1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(log2FC)

write_DEGdflist(DEGdf_WT_240_vsSerum_down2x_FDR5)

```

Then for deltaGAT201 cells

```{r DEGs_vsSerum_delGAT201}

DEGdf_delGAT201_30_vsSerum_up2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "del_30",
                log2FC > 1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(desc(log2FC))

write_DEGdflist(DEGdf_delGAT201_30_vsSerum_up2x_FDR5)

DEGdf_delGAT201_30_vsSerum_down2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "del_30",
                log2FC < -1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(log2FC)

write_DEGdflist(DEGdf_delGAT201_30_vsSerum_down2x_FDR5)


DEGdf_delGAT201_120_vsSerum_up2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "del_120",
                log2FC > 1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(desc(log2FC))

write_DEGdflist(DEGdf_delGAT201_120_vsSerum_up2x_FDR5)

DEGdf_delGAT201_120_vsSerum_down2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "del_120",
                log2FC < -1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(log2FC)

write_DEGdflist(DEGdf_delGAT201_120_vsSerum_down2x_FDR5)


DEGdf_delGAT201_240_vsSerum_up2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "del_240",
                log2FC > 1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(desc(log2FC))

write_DEGdflist(DEGdf_delGAT201_240_vsSerum_up2x_FDR5)

DEGdf_delGAT201_240_vsSerum_down2x_FDR5 <- 
  dplyr::filter(deseq_df_GT_vsserum,
                GAT201_Time == "del_240",
                log2FC < -1, padj < 0.05) %>%
  dplyr::select(-GAT201_Time) %>%
  dplyr::arrange(log2FC)

write_DEGdflist(DEGdf_delGAT201_240_vsSerum_down2x_FDR5)

```

## Notes on genes at 4h

```{r print_DEGdf_delGAT201_240_vsSerum_down2x_FDR5}
DEGdf_delGAT201_240_vsSerum_down2x_FDR5
```

- CNAG_00979, CTR4 Copper uptake transporter
- CNAG_02775, BIM1 BCS (copper-starvation) inducible membrane protein
- CNAG_00895, ZIP1 zinc transporter
- CNAG_03398, ZIP2 zinc transporter
- CNAG_03007, hypothetical protein
- CNAG_02005, hypothetical protein
- CNAG_03453, CipC-like antibiotic response protein


## Volcano plot, serum-dependent DGE in wild-type at 4 hours.

```{r volcano_WT_240_vsSerum}

volcano_WT_240_vsSerum <- 
  ggplot(data = dplyr::filter(deseq_df_GT_vsserum,
                              GAT201_Time == "WT_240"),
         aes(x = log2FC, y = -log10(padj)) ) +
  geom_point(size = 0.5, colour = "grey50") +
  geom_point(data = DEGdf_WT_240_vsSerum_up2x_FDR5,
             size = 1, colour = "darkblue") +
  geom_point(data = DEGdf_WT_240_vsSerum_down2x_FDR5,
             size = 1, colour = "darkred") +
  scale_y_continuous("-log10(p)",
                     limits = c(0,20), expand = c(0,0),
                     oob=scales::squish) +
  scale_x_continuous("log2 fold-change, RPMI 240 mins.\n← less with serum             more with serum →",
                     breaks = -5:5,
                     limits = c(-4,4), expand = c(0.01,0.01),
                     oob = scales::squish) +
  coord_cartesian(clip = "off") +
  labs(title = "GAT201 WT")

volcano_WT_240_vsSerum
```


## Volcano plot, serum-dependent DGE in gat201∆ at 4 hours.

```{r volcano_delGAT201_240_vsSerum}

volcano_delGAT201_240_vsSerum <- 
  ggplot(data = dplyr::filter(deseq_df_GT_vsserum,
                              GAT201_Time == "del_240"),
         aes(x = log2FC, y = -log10(padj)) ) +
  geom_point(size = 0.5, colour = "grey50") +
  geom_point(data = DEGdf_delGAT201_240_vsSerum_up2x_FDR5,
             size = 1, colour = "darkblue") +
  geom_point(data = DEGdf_delGAT201_240_vsSerum_down2x_FDR5,
             size = 1, colour = "darkred") +
  scale_y_continuous("-log10(p)",
                     limits = c(0,20), expand = c(0,0),
                     oob=scales::squish) +
  scale_x_continuous("log2 fold-change, RPMI 240 mins.\n← less with serum             more with serum →",
                     breaks = -5:5,
                     limits = c(-4,4), expand = c(0.01,0.01),
                     oob = scales::squish) +
  coord_cartesian(clip = "off") +
  labs(title = "gat201∆")

volcano_delGAT201_240_vsSerum
```

## Multipanel plot showing Serum affects expression of a small set of transcripts.

```{r multipanel_serumdependence, fig.width = 5, fig.height = 8}
theme_gridplot <-  
  theme_cowplot(font_size = 12, font_family = "Arial") +
  theme(panel.grid.major = element_line(colour = "grey90", size = 0.5))

multipanel_serumdependence <- 
  plot_grid(barplot_countDEGs_GT_vsserum + theme_gridplot,
            volcano_WT_240_vsSerum  + theme_gridplot,
            volcano_delGAT201_240_vsSerum  + theme_gridplot,
            ncol = 1, labels = "AUTO"
  )

multipanel_serumdependence

ggsave("../results/serum_dependence_count_volcano2x.png",
       multipanel_serumdependence,
       width = 5, height = 8)
```


# TO DO

- Write figure legend/captions.
