---
title: "Cryptococcus Gat201 RNA-seq Counts analysis"
author: "Edward Wallace"
date: "02/03/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(cowplot)
theme_set(theme_cowplot(font_size = 12) + theme(strip.background = element_blank()))
```

# Summary

This file analyses the counts per gene across samples.

* reads in the sample sheet `input_experiment/QSData_samplesheet.csv`
* reads in raw count data for all samples from file `quantseqfwd_EH_050221/counts.txt`, output by `featureCounts` running in the nextflow pipeline, which has gene (Geneid) as rows and counts as columns
* normalizes to counts per million for each sample using function `calculate_cpm`
* makes a "tidy" or long-format dataset with counts and CPM spread out by Geneid and Sampleid and sample information
* uses tidy data to plot expression against time for individual genes

This analysis has largely been superseded by rlog (regularized logarithm) normalisation.


### Define functions

Define some functions that will be useful for the analysis

```{r count_functions}
calculate_cpm <- function(x, round_digits = NA) {
    cpm <- x / sum(x, na.rm=TRUE) * 1e6
    if( !is.na(round_digits) ) {
        cpm <- round(cpm, digits = round_digits)
    }
    cpm
}
```

## Load counts and sample sheet

```{r load_counts}
sample_sheet <- 
  here::here("input_experiment","QSData_samplesheet.csv") %>%
  read_csv() %>%
  select(Sampleid = Sample.No., BioRep, Code, Strain, Condition, Time) %>%
  mutate(Gat201 = factor(Strain %in% c("A","a"), 
                         levels = c("TRUE","FALSE"),
                         labels = c("WT", "delta"))
  )

counts_bygene_wide <- 
  here::here("quantseqfwd_EH_050221","counts.txt") %>%
  read_tsv(comment = "#")
counts_bygene_wide
```

## Make tidy (long-format) counts for plotting

```{r make_tidy_counts, dependson = "load_counts"}
counts_bygene_long <- counts_bygene_wide %>%
    select(-Chr,-Start,-End,-Strand,-Length) %>%
    pivot_longer(ends_with(".bam"), names_to = "Sampleid", values_to = "Count") %>%
    mutate(Sampleid = str_extract(Sampleid, "[0-9b]*(?=_)")) %>%
    group_by(Sampleid) %>%
    mutate(TPM = calculate_cpm(Count, round_digits = 2)) %>%
    ungroup() %>%
    left_join(sample_sheet, by = "Sampleid") %>%
    select(Sampleid, Code, Strain, Gat201, Condition, Time, BioRep, Geneid, Count, TPM) %>%
    arrange(Geneid, Strain, Time, Condition, BioRep)

counts_bygene_long
```

### Write out to file `quantseqfwd_EH_050221_tidy.txt`

```{r write_tidy_counts, dependson = "make_tidy_counts"}
file_counts_long <- here::here("results",
                               "counts_quantseqfwd_EH_050221_tidy.txt.gz"
                               )

write_lines(c("# quantseqfwd_EH_050221_tidy.txt",
              "# Tidy/long-format counts from QuantSeq Gat201 RNA-seq timecourse in Cryptococcus neoformans",
              "# Edward Wallace, Edward.Wallace@ed.ac.uk",
              paste("# ", date()),
              "# Columns:",
              "#   Sampleid: id for sequencing sample (disordered here)",
              "#   Code: encodes Strain, Condition, Time",
              "#   Strain: yeast strain, a (KN99a), A (KN99alpha), B (Bahn collection deltaGat201), M (Madhani collection deltaGat201)",
              "#   Gat201: genotype for Gat201, either WT (present) or delta (deletion/disruption)",
              "#   Condition: growth media  Y (YPD), R (RPMI), RS (RPMI + serum)",
              "#   Time: time in minutes from shift to R/RS media. 0 minutes is in YPD",
              "#   BioRep: biological replicate (1 or 2)",
              "#   Geneid: Geneid with systematic Cryptococcus H99 gene name",
              "#   Count: QuantSeq 3'-end counts assigned to gene",
              "#   TPM: Transcripts per million (normalized Count for each sample)",
              "# "),
            file = file_counts_long)
write_tsv(counts_bygene_long, file = file_counts_long, col_names = TRUE, append = TRUE)
```

## Plot some individual genes


```{r plot_my_gene, dependson = "make_tidy_counts", fig.height = 3, fig.width = 3.5}
plot_my_genes_RS_line <- function(mygenes = "CNAG_06125",
                                  myconditions = c("Y","RS"),
                                  mydata = counts_bygene_long) {
    ggplot(data = filter(mydata, 
                         Geneid %in% mygenes, 
                         Condition %in% myconditions),
           aes(x = Time, y = TPM, colour = Gat201)) +
        geom_point(aes(shape = Strain)) +
        stat_summary(aes(group=Gat201),geom="line",fun = mean) +
        scale_y_log10() +
        facet_wrap(~Geneid) +
        scale_colour_manual(values = c("green4","purple"))
}
plot_my_genes_RS_line("CNAG_06125") + labs(title = "TEF1 in RPMI+Serum")
plot_my_genes_RS_line("CNAG_01551") + labs(title = "GAT201 in RPMI+Serum")
plot_my_genes_RS_line("CNAG_03012") + labs(title = "CQS1 in RPMI+Serum")
plot_my_genes_RS_line("CNAG_06346") + labs(title = "BLP1 in RPMI+Serum")
```

